### DISCUSSION TITLE
mlpyqtgraph - Matplotlib like plotting with pyqtgraph

### DISCUSSION THREAD

**Discussion by swvanbuuren** [pyqtgraph/pyqtgraph#2588] [Show and tell]
Recently I started out with a new project: [mlpyqtgraph](https://github.com/swvanbuuren/mlpyqtgraph)

**Why this project?**
I wanted to be able to plot images as easily as with [matplotlib](https://matplotlib.org/) but also keep some of the nice features of [pyqtgraph](https://github.com/pyqtgraph/pyqtgraph) such as speed and interactivity. I also wanted it to be easy to quickly integrate into an existing python program, without adding too much code. I have worked a lot with Matlab in the past, which also motivated me into doing taking this approach.

Please note that [mlpyqtgraph](https://github.com/swvanbuuren/mlpyqtgraph) is currently in a proof-of-concept stage. It lacks the majority of features you would expect and has not been optimized or polished in any way.

Before continuing with this project, I would like to sort out a few questions and get feedback:
- Does the idea make sense or is there already something similar or comparable out there? (I want to avoid reinventing the wheel)
- What do you think about my current approach? Is there something in there I should definitely change/improve?
- Are there any features that definitely should be integrated?
- Other feedback is also welcome!

I really would appreciate feedback from the [mlpyqtgraph](https://github.com/swvanbuuren/mlpyqtgraph) community and hopefully get (some of) my questions answered.

Thanks!

**Comment by j9ac9k** [comment]
Going to tag @ksunden here as this is something very much in his wheelhouse.

From what I understand, mpl is working on a typing.prototype API interface which at a glance showed promise with PyQtGraph being able to adopt as well, having very very similar APIs.  As far as I know, support hasn't been rolled out to mpl yet, which is partially why we haven't looked at adding support for it yet.

I think this is the repo for it:

https://github.com/tacaswell/data-prototype

Will check out your repo shortly!

**Reply by ksunden** [reply]
(The canonical repo has moved to https://github.com/matplotlib/data-prototype)

I've not been doing quite as much on that repo of late (I've been focusing on matplotlib/matplotlib#24976) but looking to get back to that project soon.

Data rework is certainly a fair way off still, but is one of my main focuses (likely at least 6 months).

**Comment by j9ac9k** [comment]
Hi @swvanbuuren 

Finally have a little more time to sit down, read this over, and reply.  To answer your questions a bit more directly.

> Does the idea make sense or is there already something similar or comparable out there? (I want to avoid reinventing the wheel)

As far as I know, there is nothing out there that is equivalent.

> What do you think about my current approach? Is there something in there I should definitely change/improve?

I just took a quick look through the repo, nothing stood out, I think this is going to be one of those cases if it more or less works, that's going to speak for itself.

> Are there any features that definitely should be integrated?

This one is tough for me to answer, as our two libraries often have very different use cases.  I would say, throw up some screenshots tho!!!!


I brought up the `data-prototype` repo as I think that ~~if~~ when matplotlib adopts it, we might be able to adopt it as well, making it much easier to have a more compatible API.

I guess I might ask you some questions, are there features you are having trouble implementing?

**Reply by swvanbuuren** [reply]
Some features I was thinking of:
- Option to have an iconbar at the top, to quickly access functions, that in pyqtgraph are hidden under the right-mouseclick menu
- Changing mouse cursor for different modes (pan, zoom, data-picker, etc.)
- Having a pick-and-show data point coordinates (data-picker) on a line out-of-the-box (I know there's an example "Crosshair / Mouse interaction", but that's pretty crude tbh)
- Support for more plot types. Current only line plots and basic 3D surface plots are supported.

There's more, but those are the first things that come to mind. Perhaps some of those things can also be upstreamed to `pyqtgraph`.

Speaking of 3D plots. There is a note in the help, stating that:
> pyqtgraph.opengl is based on the deprecated OpenGL fixed-function pipeline. Although it is currently a functioning system, it is likely to be superceded in the future by [VisPy](http://vispy.org/).

Does it make sense to contribute to pyqtgraph's 3D Graphics System? Or would you discourage that?

I was thinking of adding an option to show a grid with tick labels in 3D plots as well as adding functionality to plot 3D surfaces with a mesh on them. The latter is already in `mlpyqtgraph` (really a minor change, but still nice to have IMO).

**Comment by j9ac9k** [comment]
> Changing mouse cursor for different modes (pan, zoom, data-picker, etc.)

This,....we should adopt in our library, would totally merge a PR with this change.

> Support for more plot types. Current only line plots and basic 3D surface plots are supported.

Yeah, this has been a sticking point for me too, but fair warning, a _lot_ of plot types we don't support involve non-linear transformations, which will absolutely kill performance, and Qt doesn't give us anything to work with here to keep things up.  That said there is work in this area! To track progress, take a look at https://github.com/campagnola/coorx  Once this is ready for pyqtgraph integration, non-linear plots (polar plots for example) will be much more feasible.

> Does it make sense to contribute to pyqtgraph's 3D Graphics System? Or would you discourage that?

We definitely encourage contributions to pyqtgraph's 3D graphics system.  If you want to become a pyqtgraph maintainer and have opengl expertise, we want to hear from you!! So, that statement is not quite accurate but not off-base either.  PyQtGraph's 3D capability has been on bug-fix and compatibility mode primarily.  (warning, the following information may no longer be true) PyQtGraph generally has better integration capability than VisPy, think capturing mouse events, integrating keyboard events, etc etc.  VisPy will blow pyqtgraph out of the water with regards to rendering performance.  I think once VisPy has robust support for non-opengl backends (which they're working on), we will likely investigate closer integration between the two libraries.  Last I checked, they weren't ready for that yet.

If you have a change you'd like to propose, we'd certainly consider it!

**Comment by swvanbuuren** [comment]
@j9ac9k

Ok, it took me a while, but now I am back at it.

It particularly took longer, since I realized there were some fundamental issues with the underlying mechanism, that separates the thread with GUI stuff and main thread (where the normal python code runs) in the earlier version of [mlpyqtgraph](https://github.com/swvanbuuren/mlpyqtgraph).

I decided to take it apart and encapsulate the whole GUI/main thread separation mechanism into an own python package called [pqthreads](https://github.com/swvanbuuren/pqthreads). Once separated, I first worked on getting it run stable, by adding basic tests. I added PyPi releases to be able to use it in [mlpyqtgraph](https://github.com/swvanbuuren/mlpyqtgraph). And I also improved the ease of use: now you only need a single decorator to use it in another program (see the [usage section](https://github.com/swvanbuuren/pqthreads#usage) for more background information).
There are certainly more things to improve there, but for the time being, I recon it's good enough.

Now, I picked up [mlpyqtgraph](https://github.com/swvanbuuren/mlpyqtgraph) again. I really wanted to have a more decent way to display 3D data, so the first thing I added was 3D grid axis/ticks, combined with features to plot 3D surfaces and lines. The examples [surface.py](https://github.com/swvanbuuren/mlpyqtgraph/blob/master/examples/surface.py), [arctan2.py](https://github.com/swvanbuuren/mlpyqtgraph/blob/master/examples/arctan2.py) and [lorenz_attractor.py](https://github.com/swvanbuuren/mlpyqtgraph/blob/master/examples/lorenz_attractor.py) demonstrate these features. Please note, that there are still many things to be done here and it's far from perfect.

For those who don't have the time/patience to test it themselves, here are some screenshots of the new plot functionalities:
![image](https://github.com/user-attachments/assets/e62e15fb-e76d-49d8-8c7e-dce8302fbc70)
![image](https://github.com/user-attachments/assets/c4d42a7d-5526-45d5-a4cc-8d6e1baaf5a2)
![image](https://github.com/user-attachments/assets/3673c152-58a1-4af6-b8ae-ffd53a65a2ab)

Everybody here is invited to have a go and provide feedback. I'm keen to find out if the approach still makes sense.

Currently, I plan to continue to work on [mlpyqtgraph](https://github.com/swvanbuuren/mlpyqtgraph) and also upstream the 3D plot features (shown above) back to [pyqtgraph](https://github.com/pyqtgraph/pyqtgraph). I hope this would be a valuable/usable new feature.
